### PR TITLE
Feature/bigflowapi

### PR DESCRIPTION
**Pull request by gokulavasan**
Removed unwanted/unused classes.
Modified pom to generate tar.gz. More polished packaging is WIP.
Basic Clients for Standalone and Distributed Mode (launches one Flow and Ctrl-C to stop). Distributed Mode client will be enhanced (WIP) for start, stop, change instances, list flows etc.

Bamboo : https://bamboo-prod.continuuity.com/browse/PRODRELALL-UNIT1

### CODE REVIEW

**Comment by awholegunch** [comment]
Can we: change copyrights to "Copyright © 2104 Cask Data, Inc." ?
Can we add package-info.java files?

**Comment by gokulavasan** [comment]
@awholegunch Thanks for pointing those out John. I will open a PR (working on it now) right after this for the same. We want to get this in to develop soon so that we can give AT&T team a build that they can use for performance testing.

File: tigon-client/src/main/java/co/cask/tigon/DistributedMain.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  public static void finish() throws Exception {
+    if (controller != null) {
+      controller.stop().get();
+    }
+    twillRunnerService.stopAndWait();
+    metricsCollectionService.stopAndWait();
+    FileUtils.deleteDirectory(jarUnpackDir);
+  }
+
+  private static final class TestMetricsClientModule extends AbstractModule {
```
Comment (ParthGandhi) [inline tigon-client/src/main/java/co/cask/tigon/DistributedMain.java]: Remove "Test" from the name. While it currently binds to a no-op collection service, when we do implement an actual collection service we can just change the binding.

File: tigon-client/src/main/java/co/cask/tigon/DistributedMain.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+  private static Injector injector;
+  private static MetricsCollectionService metricsCollectionService;
+  private static DeployClient deployClient;
+  private static ProgramRunnerFactory programRunnerFactory;
+  private static ProgramController controller;
+  private static TwillRunnerService twillRunnerService;
+  private static CountDownLatch runLatch;
+
+  static void usage(boolean error) {
+    // Which output stream should we use?
```
Comment (ParthGandhi) [inline tigon-client/src/main/java/co/cask/tigon/DistributedMain.java]: What you have here looks okay to me.

File: tigon-client/src/main/java/co/cask/tigon/DistributedMain.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+  private static void deploy(File jarPath, String classToLoad) throws Exception {
+    Location deployedJar = deployClient.createJar(jarPath, classToLoad, jarUnpackDir);
+    Program program = Programs.createWithUnpack(deployedJar, jarUnpackDir);
+    controller = programRunnerFactory.create(ProgramRunnerFactory.Type.FLOW).run(
+      program, new SimpleProgramOptions(program.getName(), new BasicArguments(), new BasicArguments()));
+  }
+
+  public static void main(String[] args) {
+    System.out.println("Tigon Distributed");
```
Comment (ParthGandhi) [inline tigon-client/src/main/java/co/cask/tigon/DistributedMain.java]: Something a tad more descriptive would be better.  "Tigon Distributed Client" perhaps?

File: tigon-client/src/main/java/co/cask/tigon/DistributedMain.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+  private static DeployClient deployClient;
+  private static ProgramRunnerFactory programRunnerFactory;
+  private static ProgramController controller;
+  private static TwillRunnerService twillRunnerService;
+  private static CountDownLatch runLatch;
+
+  static void usage(boolean error) {
+    // Which output stream should we use?
+    PrintStream out = (error ? System.err : System.out);
+    out.println("Usage: ./run_standalone.sh <path-to-JAR> <FlowClassName>");
```
Comment (ParthGandhi) [inline tigon-client/src/main/java/co/cask/tigon/DistributedMain.java]: You mean `./run_distributed.sh` here. And on the next line.

File: tigon-client/src/main/java/co/cask/tigon/StandaloneMain.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  private static LocationFactory locationFactory;
+  private static DeployClient deployClient;
+  private static ProgramRunnerFactory programRunnerFactory;
+  private static ProgramController controller;
+  private static CountDownLatch runLatch;
+
+  static void usage(boolean error) {
+    // Which output stream should we use?
+    PrintStream out = (error ? System.err : System.out);
+    out.println("./run_distributed.sh <path-to-JAR> <FlowClassName>");
```
Comment (ParthGandhi) [inline tigon-client/src/main/java/co/cask/tigon/StandaloneMain.java]: And `./run_standalone.sh` here.

File: tigon-common/src/main/java/co/cask/tigon/conf/Constants.java [hunk 6]
```diff
@@ -53,32 +38,10 @@
...
-   */
-  public static final class Dangerous {
-    public static final String UNRECOVERABLE_RESET = "enable.unrecoverable.reset";
-    public static final boolean DEFAULT_UNRECOVERABLE_RESET = false;
-  }
-
-  /**
    * App Fabric Configuration.
    */
   public static final class AppFabric {
```
Comment (ParthGandhi) [inline tigon-common/src/main/java/co/cask/tigon/conf/Constants.java]: AppFabric? Rename to something relevant.

File: tigon-common/src/main/java/co/cask/tigon/guice/LocationRuntimeModule.java [hunk 7]
```diff
@@ -71,6 +71,7 @@ private LocalLocationFactory providesLocalLocationFactory(CConfiguration cConf)
 
   private static final class HDFSLocationModule extends AbstractModule {
 
+
```
Comment (ParthGandhi) [inline tigon-common/src/main/java/co/cask/tigon/guice/LocationRuntimeModule.java]: Extra line.

File: .gitignore [hunk 8]
```diff
@@ -30,7 +30,7 @@ hs_err_pid*
 .idea
 
 # Maven build directory
-target
+target/
```
Comment (achalpandey) [inline .gitignore:33]: can we leave it as is, so that it is consistent with the rest of the file

File: tigon-client/pom.xml [hunk 9]
```diff
@@ -26,5 +26,23 @@ the License.
...
+    <dependency>
+      <groupId>org.apache.twill</groupId>
+      <artifactId>twill-common</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>com.google.guava</groupId>
+      <artifactId>guava</artifactId>
+    </dependency>
+  </dependencies>
+</project>
```
Comment (achalpandey) [inline tigon-client/pom.xml]: Add new line?

File: tigon-common/src/main/java/co/cask/tigon/guice/TwillModule.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+import org.apache.hadoop.yarn.conf.YarnConfiguration;
+import org.apache.twill.api.TwillRunner;
+import org.apache.twill.api.TwillRunnerService;
+import org.apache.twill.filesystem.LocationFactories;
+import org.apache.twill.filesystem.LocationFactory;
+import org.apache.twill.yarn.YarnTwillRunnerService;
+
+/**
+ * Guice module for providing bindings for Twill. This module requires accessible bindings to
+ * {@link CConfiguration}, {@link org.apache.hadoop.yarn.conf.YarnConfiguration} and
```
Comment (ParthGandhi) [inline tigon-common/src/main/java/co/cask/tigon/guice/TwillModule.java:34]: Fix your IntelliJ to not use a fully qualified classname. 

Preferences -> Code Style -> Java -> Imports Tab -> Uncheck "Use fully qualified class names in javadoc". 

Thanks to @rohitsinha54  for the tip.

File: tigon-common/src/main/java/co/cask/tigon/guice/TwillModule.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    bind(TwillRunner.class).to(TwillRunnerService.class);
+
+    expose(TwillRunnerService.class);
+    expose(TwillRunner.class);
+  }
+
+  /**
+   * Provider method for instantiating {@link org.apache.twill.yarn.YarnTwillRunnerService}.
+   */
+  @Singleton
```
Comment (ParthGandhi) [inline tigon-common/src/main/java/co/cask/tigon/guice/TwillModule.java:51]: Does this need to be annotated with `@Singleton` if the binding above is already in `Scopes.SINGLETON`? 

(genuine question. I don't know the answer)

File: tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/AbstractMasterTwillRunnable.java [hunk 12]
```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2014 Cask Data, Inc.
```
Comment (ParthGandhi) [inline tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/AbstractMasterTwillRunnable.java]: All the new files here are using an outdated copyright. I presume other files in the rest of the codebase are doing the same. 

For now its okay, but lets make a note to go over the entire codebase once to ensure we're using the correct version. 

This is the latest version: https://wiki.continuuity.com/display/ENG/Setup+Environment

File: tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/AbstractMasterTwillRunnable.java [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+    } catch (InterruptedException e) {
+      LOG.debug("Waiting on latch interrupted {}", name);
+      Thread.currentThread().interrupt();
+    } catch (ExecutionException e) {
+      LOG.error("Exception in service.", e);
+      throw Throwables.propagate(e);
+    }
+
+    List<Service> reverse = Lists.reverse(services);
+    Services.chainStop(reverse.get(0), reverse.subList(1, reverse.size()).toArray(new Service[0]));
```
Comment (ParthGandhi) [inline tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/AbstractMasterTwillRunnable.java]: This should happen in `destroy()`, right?

File: tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/AbstractMasterTwillRunnable.java [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+      runThread.interrupt();
+    }
+  }
+
+  /**
+   * Class extending AbstractReactorTwillRunnable should populate services
+   * with a list of Services which will be started in increasing order of index.
+   * The services will be stopped in the reverse order.
+   */
+  protected abstract void getServices(List<? super Service> services);
```
Comment (ParthGandhi) [inline tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/AbstractMasterTwillRunnable.java]: Either rename this method to `setServices` or have it return a list of services that can be set in here.

File: tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/TransactionServiceTwillRunnable.java [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+  @Override
+  protected void doInit(TwillContext context) {
+    try {
+      getCConfiguration().set(Constants.Transaction.Container.ADDRESS, context.getHost().getCanonicalHostName());
+
+      Injector injector = createGuiceInjector(getCConfiguration(), getConfiguration());
+
+      LOG.info("Initializing runnable {}", name);
+      // Set the hostname of the machine so that cConf can be used to start internal services
+      LOG.info("{} Setting host name to {}", name, context.getHost().getCanonicalHostName());
```
Comment (ParthGandhi) [inline tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/TransactionServiceTwillRunnable.java]: Can you move line 68 to just above this LOG. Helps readability since ideally you want to log when you do the action.

File: tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/TransactionServiceTwillRunnable.java [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+      new MetricsClientRuntimeModule().getDistributedModules()
+    );
+  }
+
+  private static Module createDataFabricModule() {
+    return Modules.override(new DataFabricModules().getDistributedModules()).with(new AbstractModule() {
+      @Override
+      protected void configure() {
+        // Bind to provider that create new instances of storage and tx manager every time.
+        bind(TransactionStateStorage.class).annotatedWith(Names.named("persist"))
```
Comment (ParthGandhi) [inline tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/TransactionServiceTwillRunnable.java:110]: Unused binding.

**Comment by ParthGandhi** [comment]
As discussed offline, also consider moving `TestBase` and the relevant managers to an appropriate place.

File: .gitignore (hunk 8 above)
Comment (gokulavasan) [inline .gitignore:33]: Actually the way to ignore directories is to have a "/" at the end. http://git-scm.com/docs/gitignore

File: tigon-client/src/main/java/co/cask/tigon/DistributedMain.java (hunk 1 above)
Comment (gokulavasan) [inline tigon-client/src/main/java/co/cask/tigon/DistributedMain.java]: changed.

File: tigon-client/src/main/java/co/cask/tigon/DistributedMain.java (hunk 2 above)
Comment (gokulavasan) [inline tigon-client/src/main/java/co/cask/tigon/DistributedMain.java]: removed. that was a comment in cdap though. we should remove it there too?

File: tigon-client/src/main/java/co/cask/tigon/DistributedMain.java (hunk 3 above)
Comment (gokulavasan) [inline tigon-client/src/main/java/co/cask/tigon/DistributedMain.java]: Changed.

File: tigon-client/src/main/java/co/cask/tigon/DistributedMain.java (hunk 4 above)
Comment (gokulavasan) [inline tigon-client/src/main/java/co/cask/tigon/DistributedMain.java]: Yes! Changed.

File: tigon-common/src/main/java/co/cask/tigon/conf/Constants.java (hunk 6 above)
Comment (gokulavasan) [inline tigon-common/src/main/java/co/cask/tigon/conf/Constants.java]: Removed more constants. Removed.

File: tigon-common/src/main/java/co/cask/tigon/guice/TwillModule.java (hunk 10 above)
Comment (gokulavasan) [inline tigon-common/src/main/java/co/cask/tigon/guice/TwillModule.java:34]: I have marked this in my IDE. Will send another PR when we change license etc.

File: tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/AbstractMasterTwillRunnable.java (hunk 12 above)
Comment (gokulavasan) [inline tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/AbstractMasterTwillRunnable.java]: As discussed, copyright changes will be another PR.

File: tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/TransactionServiceTwillRunnable.java (hunk 15 above)
Comment (gokulavasan) [inline tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/TransactionServiceTwillRunnable.java]: Removed them since they are not required.

File: tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/AbstractMasterTwillRunnable.java (hunk 13 above)
Comment (gokulavasan) [inline tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/AbstractMasterTwillRunnable.java]: I guess this covers the case when one of the individual services throws an exception or completes? This is from CDAP.

File: tigon-common/src/main/java/co/cask/tigon/guice/TwillModule.java (hunk 11 above)
Comment (gokulavasan) [inline tigon-common/src/main/java/co/cask/tigon/guice/TwillModule.java:51]: Not super familiar. This is from CDAP. We can find out more about why we have structured it this from @chtyim

File: tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/AbstractMasterTwillRunnable.java (hunk 14 above)
Comment (gokulavasan) [inline tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/AbstractMasterTwillRunnable.java]: This is from CDAP. This is getServices from AbstractMasterTwillRunnable's pt of view. Lets talk in person.

File: tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/TransactionServiceTwillRunnable.java (hunk 16 above)
Comment (gokulavasan) [inline tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/TransactionServiceTwillRunnable.java:110]: I think this is used. CDAP has the same code. We can confirm with @chtyim

**Comment by gokulavasan** [comment]
@ParthGandhi Addressed your comments. Also moved the TestBase and managers as discussed. Please take a look.

File: .gitignore (hunk 8 above)
Comment (achalpandey) [inline .gitignore:33]: Cool, then let us fix all the directories to end in a "/"

File: src/assembly/tgz.xml [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+      <binaries>
+        <outputDirectory>lib</outputDirectory>
+        <unpack>false</unpack>
+        <dependencySets>
+          <dependencySet/>
+        </dependencySets>
+      </binaries>
+    </moduleSet>
+  </moduleSets>
+</assembly>
```
Comment (achalpandey) [inline src/assembly/tgz.xml]: Not sure, but don't we need a newline at the end of the file? wouldn't this be a mvn build error?

File: tigon-flow/src/main/java/co/cask/tigon/app/guice/DistributedProgramRunnerModule.java [hunk 18]
```diff
@@ -30,7 +30,7 @@
 /**
  * Guice module for distributed AppFabric. Used by the app-fabric server, not for distributed containers.
```
Comment (achalpandey) [inline tigon-flow/src/main/java/co/cask/tigon/app/guice/DistributedProgramRunnerModule.java]: We might want to get rid of AppFabric and other CDAP references in our code base. 

Do you think we should address this in another PR?

File: src/assembly/tgz.xml (hunk 17 above)
Comment (gokulavasan) [inline src/assembly/tgz.xml]: Newline added. Not sure but mvn is not complaining.

File: tigon-flow/src/main/java/co/cask/tigon/app/guice/DistributedProgramRunnerModule.java (hunk 18 above)
Comment (gokulavasan) [inline tigon-flow/src/main/java/co/cask/tigon/app/guice/DistributedProgramRunnerModule.java]: Changed. We need to do few rounds of cleanup but since you pointed this out, I have fixed it.

File: .gitignore (hunk 8 above)
Comment (gokulavasan) [inline .gitignore:33]: As far as I saw there is only .idea directory and I have changed that.

**Comment by gokulavasan** [comment]
@achalpandey @ParthGandhi Addressed your comments.

File: .gitignore (hunk 8 above)
Comment (achalpandey) [inline .gitignore:33]: 2 more directories :

Under 
`# Generated binaries`
`tigon-sql/bin`
`tigon-sql/lib`

**Comment by achalpandey** [comment]
One small comment, after that LGTM :+1:

File: .gitignore (hunk 8 above)
Comment (gokulavasan) [inline .gitignore:33]: Oh ok. Done.

File: tigon-client/src/main/java/co/cask/tigon/DistributedMain.java [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+import java.io.File;
+import java.io.PrintStream;
+import java.util.concurrent.CountDownLatch;
+
+/**
+ * Tigon Distributed Main.
+ */
+public class DistributedMain {
+  private static final Logger LOG = LoggerFactory.getLogger(DistributedMain.class);
+  private static final File jarUnpackDir = Files.createTempDir();
```
Comment (chtyim) [inline tigon-client/src/main/java/co/cask/tigon/DistributedMain.java]: Don't create directory in static variable, as there is no good way to make sure it will be deleted.

File: tigon-client/src/main/java/co/cask/tigon/DistributedMain.java [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+  private static MetricsCollectionService metricsCollectionService;
+  private static DeployClient deployClient;
+  private static ProgramRunnerFactory programRunnerFactory;
+  private static ProgramController controller;
+  private static TwillRunnerService twillRunnerService;
+  private static CountDownLatch runLatch;
+
+  static void usage(boolean error) {
+    PrintStream out = (error ? System.err : System.out);
+    out.println("Usage: ./run_distributed.sh <path-to-JAR> <FlowClassName>");
```
Comment (chtyim) [inline tigon-client/src/main/java/co/cask/tigon/DistributedMain.java]: This class shouldn't know about the shell script, right? Seems like printing the shell command is wrong. This help message should be printed by the shell script. While for this class, you should print something like `java DistributedMain ....`

File: tigon-client/src/main/java/co/cask/tigon/DistributedMain.java [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+import java.util.concurrent.CountDownLatch;
+
+/**
+ * Tigon Distributed Main.
+ */
+public class DistributedMain {
+  private static final Logger LOG = LoggerFactory.getLogger(DistributedMain.class);
+  private static final File jarUnpackDir = Files.createTempDir();
+
+  private static Injector injector;
```
Comment (chtyim) [inline tigon-client/src/main/java/co/cask/tigon/DistributedMain.java]: Usually I prefer non-static to static if it is not pure utility type method (e.g. you write it like a regular class, just that main() will new that class and call methods on it). It helps the design better (e.g. you can have final fields rather than bunch of non-final static fields). Also, in general, this make the class more reusable.

Comment (gokulavasan) [inline tigon-client/src/main/java/co/cask/tigon/DistributedMain.java]: Thanks for the comments Terence. Incorporated your comments.

**Comment by gokulavasan** [comment]
@chtyim Addressed your comments. Please take a look.

File: tigon-client/src/main/java/co/cask/tigon/DistributedMain.java [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+    cConf.set(Constants.CFG_LOCAL_DATA_DIR, localDataDir.getAbsolutePath());
+    Configuration hConf = HBaseConfiguration.create();
+
+    Injector injector = Guice.createInjector(createModules(cConf, hConf));
+    metricsCollectionService = injector.getInstance(MetricsCollectionService.class);
+    twillRunnerService = injector.getInstance(TwillRunnerService.class);
+    deployClient = injector.getInstance(DeployClient.class);
+  }
+
+  static void usage(boolean error) {
```
Comment (sreevatsanraman) [inline tigon-client/src/main/java/co/cask/tigon/DistributedMain.java]: why is this package private?

File: tigon-client/src/main/java/co/cask/tigon/StandaloneMain.java [hunk 23]
```diff
@@ -0,0 +1,171 @@
...
+      }
+
+      if (args.length != 2) {
+        usage(true);
+      }
+
+      File jarPath = new File(args[0]);
+      String mainClassName = args[1];
+
+      StandaloneMain main = null;
```
Comment (sreevatsanraman) [inline tigon-client/src/main/java/co/cask/tigon/StandaloneMain.java]: We can move this inside the try block. No need to initialize with null

File: tigon-client/src/main/java/co/cask/tigon/StandaloneMain.java [hunk 24]
```diff
@@ -0,0 +1,171 @@
...
+      try {
+        main = createStandaloneMain();
+        main.startUp(jarPath, mainClassName);
+      } catch (Exception e) {
+        e.printStackTrace();
```
Comment (sreevatsanraman) [inline tigon-client/src/main/java/co/cask/tigon/StandaloneMain.java]: Log the error to stderr

File: tigon-client/src/main/java/co/cask/tigon/DistributedMain.java [hunk 25]
```diff
@@ -0,0 +1,179 @@
...
+      }
+
+      if (args.length != 2) {
+        usage(true);
+      }
+
+      File jarPath = new File(args[0]);
+      String mainClassName = args[1];
+
+      DistributedMain main = null;
```
Comment (sreevatsanraman) [inline tigon-client/src/main/java/co/cask/tigon/DistributedMain.java]: Move this inside try block

File: tigon-client/src/main/java/co/cask/tigon/StandaloneMain.java [hunk 26]
```diff
@@ -0,0 +1,171 @@
...
+    cConf.set(Constants.CFG_LOCAL_DATA_DIR, localDataDir.getAbsolutePath());
+    Configuration hConf = new Configuration();
+
+    Injector injector = Guice.createInjector(createModules(cConf, hConf));
+    txService = injector.getInstance(TransactionManager.class);
+    metricsCollectionService = injector.getInstance(MetricsCollectionService.class);
+    deployClient = injector.getInstance(DeployClient.class);
+  }
+
+  static void usage(boolean error) {
```
Comment (sreevatsanraman) [inline tigon-client/src/main/java/co/cask/tigon/StandaloneMain.java]: private?

File: tigon-common/pom.xml [hunk 27]
```diff
@@ -66,6 +66,11 @@ the License.
       <groupId>org.apache.httpcomponents</groupId>
       <artifactId>httpclient</artifactId>
     </dependency>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <scope>compile</scope>
```
Comment (sreevatsanraman) [inline tigon-common/pom.xml]: scope should be test

File: src/assembly/tgz.xml [hunk 28]
```diff
@@ -0,0 +1,40 @@
+<!--
+  Copyright 2014 Cask Data, Inc.
```
Comment (sreevatsanraman) [inline src/assembly/tgz.xml]: Copy right should be 

Copyright  © 2014 Cask Data, Inc

Fix all instances.

File: tigon-common/src/main/resources/cdap-default.xml [hunk 29]
```diff
@@ -25,7 +25,7 @@
 
     <property>
         <name>zookeeper.quorum</name>
-        <value>127.0.0.1:2181/${root.namespace}</value>
+        <value>127.0.0.1/${root.namespace}</value>
```
Comment (sreevatsanraman) [inline tigon-common/src/main/resources/cdap-default.xml]: The file should not reference cdap -  tigon-default.xml?

File: tigon-common/src/main/resources/cdap-site.xml [hunk 30]
```diff
@@ -15,15 +15,6 @@
   -->
 <configuration>
```
Comment (sreevatsanraman) [inline tigon-common/src/main/resources/cdap-site.xml:16]: Same here. File name fix.

File: tigon-distribution/pom.xml [hunk 31]
```diff
@@ -0,0 +1,77 @@
...
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>tigon-distribution</artifactId>
+  <packaging>pom</packaging>
+
+  <dependencies>
+    <dependency>
+      <groupId>co.cask.tigon</groupId>
+      <artifactId>tigon-api</artifactId>
+      <version>${project.version}</version>
```
Comment (sreevatsanraman) [inline tigon-distribution/pom.xml:35]: the version should come from the parent pom. You shouldn't have to specify again.

File: tigon-distribution/pom.xml [hunk 32]
```diff
@@ -0,0 +1,77 @@
...
+                <descriptor>src/assembly/tgz.xml</descriptor>
+              </descriptors>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
+  </build>
+
+</project>
```
Comment (sreevatsanraman) [inline tigon-distribution/pom.xml]: new line

File: tigon-flow/src/main/java/co/cask/tigon/flow/DeployClient.java [hunk 33]
```diff
@@ -70,7 +91,52 @@ public static Manifest getManifestWithMainClass(Class<?> klass) {
     return manifest;
   }
 
-  public Location deployFlow(final String applicationId, Class<?> flowClz, File... bundleEmbeddedJars)
+  private static void expandJar(File jarPath, File unpackDir) throws Exception {
```
Comment (sreevatsanraman) [inline tigon-flow/src/main/java/co/cask/tigon/flow/DeployClient.java:95]: Need to handle exceptions and close the input and output streams in finally blocks.

``` java
InputStream is = jar.getInputStream(file);
FileOutputStream fos = new FileOutputStream(f);
try {  
  while (is.available() > 0) {
    fos.write(is.read());
   }
 } catch (IOException e) {
   throw Throwables.propagate(e);
 } finally {
   fos.close();
   is.close();
}
```

File: tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/AbstractMasterTwillRunnable.java [hunk 34]
```diff
@@ -0,0 +1,176 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Abstract TwillRunnable class for Master system service.
```
Comment (sreevatsanraman) [inline tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/AbstractMasterTwillRunnable.java]: What is a Master System service in Tigon?

File: tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/AbstractMasterTwillRunnable.java [hunk 35]
```diff
@@ -0,0 +1,176 @@
...
+ */
+public abstract class AbstractMasterTwillRunnable extends AbstractTwillRunnable {
```
Comment (sreevatsanraman) [inline tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/AbstractMasterTwillRunnable.java]: Find a suitable name for this class.

**Comment by gokulavasan** [comment]
@sreevatsanraman Addressed your comments. Please take a look again.

**Comment by sreevatsanraman** [comment]
LGTM :+1:
